bigwig_loader/collection.py

```diff
@@ -75,6 +73,27 @@ def run_indexing(self) -> None:
     def __len__(self) -> int:
         return len(self.bigwigs)
 
+    def reset_gpu(self) -> None:
+        """
+        Remove all gpu arrays from the previously used device and recreate when necessary on
+        current device. This is useful when training is done on multiple gpus and the arrays
+        need to be recreated on the new gpu.
+        """
+
+        self._out = cp.zeros((len(self), 1, 1), dtype=cp.float32)
+        self._decoder = None
+        self._memory_bank = None
+
+    def _get_decoder(self) -> Decoder:
+        if self._decoder:
+            return self._decoder
+        self._decoder = Decoder(
+            max_rows_per_chunk=self.max_rows_per_chunk,
+            max_uncompressed_chunk_size=self.max_rows_per_chunk * 12 + 24,
+            chromosome_offsets=self.local_chrom_ids_to_offset_matrix,
+        )
+        return self._decoder
```

Comment:
You could use a cached_property here to simplify this a bit.

You'd just need to switch your code in reset_gpu to delete from `__dict__`

Reply:
So, I am mostly using this pattern because mypy tends to complain when I set (and type) self._decoder in init and then have a property:

```python
@property
def decoder(self):
    ...
```

I can be wrong. Does cached_property fix exactly that problem?

Reply:
yep.

You get rid of `self._decoder` and you just have:

```python
@cached_property
def decoder(self) -> Decoder:
    return Decoder(...)
```

Clearing the cache looks like this:

```python
def clear(self):
    if "decoder" in self.__dict__:
        del self.__dict__["decoder"]
```

It's useful to do it like that if all you want is late initialization of self.decoder